### DISCUSSION TITLE
motr example could not initial when "struct m0_config" not global variable

### DISCUSSION THREAD

**Discussion by kiwionly2** [Seagate/cortx#1482] [General]
Hi, 

I was playing around example from `cortx-motr/motr/examples`. 

For `example1.c`, when I moving `struct m0_config` from global variable to local variable close to where it use, it compile successfully, however fail to run.

```
static struct m0_container       motr_container;
// static struct m0_config          motr_conf;   // comment the code at top
static struct m0_idx_dix_config  motr_dix_conf;
```

```
	motr_dix_conf.kc_create_meta = false;

	struct m0_config motr_conf; // move m0_config to local

	motr_conf.mc_is_oostore            = true;
	motr_conf.mc_is_read_verify        = false;
	motr_conf.mc_ha_addr               = argv[1];
	motr_conf.mc_local_addr            = argv[2];
	motr_conf.mc_profile               = argv[3];
	motr_conf.mc_process_fid           = argv[4];
	motr_conf.mc_tm_recv_queue_min_len = M0_NET_TM_RECV_QUEUE_DEF_LEN;
	motr_conf.mc_max_rpc_msg_size      = M0_RPC_DEF_MAX_RPC_MSG_SIZE;
	motr_conf.mc_idx_service_id        = M0_IDX_DIX;
	motr_conf.mc_idx_service_conf      = (void *)&motr_dix_conf;

	rc = m0_client_init(&m0_instance, &motr_conf, true);
	if (rc != 0) {
		printf("error in m0_client_init: %d\n", rc);
		exit(rc);
	}
```

The error as below ( look like it read some uninitialize value )

```
...
bash-4.2$ gcc -I/home/502380/cortx-motr -DM0_EXTERN=extern -DM0_INTERNAL= -Wno-attributes -L/home/502380/cortx-motr/motr/.libs -lmotr example1.c -o example1
bash-4.2$ ./example1 192.168.63.52@tcp:12345:34:1 192.168.63.52@tcp:12345:33:1001 "<0x7000000000000001:0>" "<0x7200000000000001:64>" 1234567567
motr[12544]:  e290   WARN  [client_init.c:1687:m0_client_init]  ADDB size is more than recommended
motr[12544]:  df80  FATAL  [lib/assert.c:50:m0_panic]  panic: (stor_rounded(stor, size)) at m0_addb2_storage_init() (addb2/storage.c:386)  [git: 2.0.0-585-80-ge4f9afe6] /home/502380/cortx-motr/motr/examples/m0trace.12544
Motr panic: (stor_rounded(stor, size)) at m0_addb2_storage_init() addb2/storage.c:386 (errno: 2) (last failed: none) [git: 2.0.0-585-80-ge4f9afe6] pid: 12544  /home/502380/cortx-motr/motr/examples/m0trace.12544
/home/502380/cortx-motr/motr/.libs/libmotr.so.2(m0_arch_backtrace+0x20)[0x7f8b255f5d40]
/home/502380/cortx-motr/motr/.libs/libmotr.so.2(m0_arch_panic+0xe6)[0x7f8b255f5ef6]
/home/502380/cortx-motr/motr/.libs/libmotr.so.2(+0x3642c4)[0x7f8b255e62c4]
...
```

Since `struct m0_config` only use by `m0_client_init`, shouldn't it suppose to work at local scope ?  

One of the advantage of local scope is it will enabled developer create multiple client instance by encapsulate client variable ( global variable in this example) to a struct as below:

```
// we could create multiple client connection to motr
typedef struct MotrClient {
	struct m0_client    	      *instance;
	struct m0_container      container;
       struct m0_config            motr_conf
} MotrClient;

```

*not sure if this is a bug, I open at discussion instead.

**Comment by huanghua78** [comment]
This is wrong.

>For example1.c, when I moving struct m0_config from global variable to local variable close to where it use, it compile successfully, 
> however fail to run.


Please read the code of m0_client_init().
It saves the `conf` pointer in m0_client and uses it later. So, the `motr_conf` must be global or be dynamically allocated.

>One of the advantage of local scope is it will enabled developer create multiple client instance by encapsulate client variable ( global variable in this example) to a struct as below:
>

Motr uses many global variables per-process. In one process, only one single Motr client process is supported.

**Reply by kiwionly2** [reply]
@huanghua78  

Thank for the explanation and clarification, 

It is hard to find those information unless have a deep understanding of underlying  architecture.

**Comment by huanghua78** [comment]
@kiwionly2 If you want to have multi-threaded Motr applications,
please refer to:

motr/st/utils/cc_cp_cat.c
